### DISCUSSION TITLE
Tox 4.x: specify multiline lists in variables? (or expand them without the newline characters)

### DISCUSSION THREAD

**Discussion by ppentchev** [tox-dev/tox#2922] [Q&A] [answered]
## Issue

Hi,

Thanks a lot for writing, maintaining, and developing Tox!

So I'm not sure that this is a bug report as such, since, as noted below, I fully understand why Tox does what it does and why. It's more of a "is there a way to configure Tox 4.x to do something similar to what I did in Tox 3.x" question.

In short, I have something similar to the following setup in lots of my Tox 3.x tox.ini files (I added "skipsdist" for this pared-down example):
```ini
[tox]
envlist = mypy,pylint
skipsdist = True

[defs]
pyfiles =
  src/confget
  src/some_other_tool_or_module
  unit_tests

[testenv:mypy]
deps =
  mypy >= 1, < 2
commands =
  mypy -- {[defs]pyfiles}

[testenv:pylint]
deps =
  pylint >= 2.16, < 2.17
commands =
  pylint -- {[defs]pyfiles}
```

So this works with Tox 3.x: it expands the variable, apparently replaces the newline characters with spaces or something like that, and then runs the command on all the files. The same tox.ini syntax with Tox 4.x leads to Tox passing `pylint -- src/confget<newline>  src/...</newline>  unit_tests` to the shell, and so the shell treats those as three separate commands, which is not exactly the intent here :slight_smile: 

I mean, I understand why Tox 4.x would do that: after all, I *did* include some newline characters in the variable's value, so it wants to reflect that in the command. Still, is there a way to achieve what I want in Tox 4.x syntax? I know I can specify all the files on a single line, but I kind of like the readability of the one-per-line format...

Thanks again, and keep up the great work!

## Environment

This can be reproduced in a minimal virtual environment that only contains tox, e.g. one obtained by `nix-shell --pure -p 'python3.withPackages(p: [p.pip p.tox])'` which, on my Debian system with the Nix unstable channel, results in the following list of packages:

```console
Package       Version
------------- -------
cachetools    5.2.1
chardet       5.1.0
colorama      0.4.6
distlib       0.3.6
filelock      3.9.0
packaging     21.3
pip           22.3.1
platformdirs  2.6.2
pluggy        1.0.0
py            1.11.0
pyparsing     3.0.9
pyproject_api 1.2.1
six           1.16.0
toml          0.10.2
tomli         2.0.1
tox           4.1.0
virtualenv    20.17.1
```
...and empty `__init__.py` files in the `src/confget`, `src/some_other_tool_or_module`, and `unit_tests` subdirectories.

## Output of running tox

This is the output of `tox -r`, I don't think the verbose flags are really needed in this specific case:

```console
[nix-shell:~/lang/python/misc/tox-newlines]$ tox -r
mypy: remove tox env folder /home/roam/lang/python/misc/tox-newlines/.tox/mypy
mypy: install_deps> python -I -m pip install 'mypy<2,>=1'
mypy: commands[0]> mypy -- src/confget
Success: no issues found in 1 source file
mypy: commands[1]> src/some_other_tool_or_module
mypy: exit 2 (0.00 seconds) /home/roam/lang/python/misc/tox-newlines> src/some_other_tool_or_module
mypy: FAIL ✖ in 2.16 seconds
pylint: remove tox env folder /home/roam/lang/python/misc/tox-newlines/.tox/pylint
pylint: install_deps> python -I -m pip install 'pylint<2.17,>=2.16'
pylint: commands[0]> pylint -- src/confget
pylint: commands[1]> src/some_other_tool_or_module
pylint: exit 2 (0.00 seconds) /home/roam/lang/python/misc/tox-newlines> src/some_other_tool_or_module
  mypy: FAIL code 2 (2.16=setup[2.05]+cmd[0.10,0.00] seconds)
  pylint: FAIL code 2 (2.35=setup[1.97]+cmd[0.38,0.00] seconds)
  evaluation failed :( (4.54 seconds)

[nix-shell:~/lang/python/misc/tox-newlines]$
```

## Minimal example

This is the `tox.ini` file (the same as in the "Issue" section above):

```console
[roam@straylight ~/lang/python/misc/tox-newlines]$ cat > tox.ini <<EOINI
[tox]
envlist = mypy,pylint
skipsdist = True

[defs]
pyfiles =
  src/confget
  src/some_other_tool_or_module
  unit_tests

[testenv:mypy]
deps =
  mypy >= 1, < 2
commands =
  mypy -- {[defs]pyfiles}

[testenv:pylint]
deps =
  pylint >= 2.16, < 2.17
commands =
  pylint -- {[defs]pyfiles}
EOINI
[roam@straylight ~/lang/python/misc/tox-newlines]$ mkdir -p src/confget src/some_other_tool_or_module unit_tests
[roam@straylight ~/lang/python/misc/tox-newlines]$ touch src/confget/__init__.py src/some_other_tool_or_module/__init__.py unit_tests/__init__.py
[roam@straylight ~/lang/python/misc/tox-newlines]$
```

**Comment by gaborbernat** [comment, marked as answer]
Can't you `\` to specify the lines should be merged?

```ini
[defs]
pyfiles =
  src/confget \
  src/some_other_tool_or_module \
  unit_tests
```

**Reply by ppentchev** [reply]
Ah, I forgot that I had tried that, and forgot to mention it. Actually, yes, this does work in Tox 4.x, but then...
```console
[roam@straylight ~/lang/python/misc/tox-newlines]$ tox --version
3.28.0 imported from /usr/lib/python3/dist-packages/tox/__init__.py
[roam@straylight ~/lang/python/misc/tox-newlines]$ tox -r -e mypy
mypy recreate: /home/roam/lang/python/misc/tox-newlines/.tox/mypy
mypy installdeps: mypy >= 1, < 2
mypy installed: mypy==1.0.0,mypy-extensions==1.0.0,typing_extensions==4.4.0
mypy run-test-pre: PYTHONHASHSEED='723939847'
mypy run-test: commands[0] | mypy -- src/confget '\' src/some_other_tool_or_module '\' unit_tests
mypy: can't read file '\': No such file or directory
ERROR: InvocationError for command /home/roam/lang/python/misc/tox-newlines/.tox/mypy/bin/mypy -- src/confget '\' src/some_other_tool_or_module '\' unit_tests (exited with code 2)
_______________________________________________________ summary _______________________________________________________
ERROR:   mypy: commands failed
[roam@straylight ~/lang/python/misc/tox-newlines]$
```

So hm... I wonder if there is a way to make things conditional on the Tox version... or just straight away declare that the projects can now only use Tox 4.x? Hmm, but there are some OS's and distributions where Tox 4.x has not been packaged yet...

OK, thanks, your suggestion does help with Tox 4.x. Thanks! That's a way forward, at least. Now would it be too annoying of me to ask whether you can think of a way to make it work with Tox 3.x, too? :)

Thanks again, and sorry for all the mess I created today with not being able to figure out how to ask a simple question...

G'luck,
Peter

**Reply by jugmac00** [reply]
Afaik it is not possible to have an entirely tox 3 and 4 compatible configuration file for a couple of edge cases.

Please note that tox 3 is no longer supported by us, but maybe by linux distributions.

You can possibly workaround the compatibility issues by using the `min_version` directive:
https://tox.wiki/en/latest/config.html#min_version

You could add tox 4 as a minimum, and when you then run it with tox 3, tox will bootstrap itself in version 4.
(Disclaimer I have never used that and I am not entirely sure whether that works, but it should).

**Reply by gaborbernat** [reply]
We have no intention of keeping a configuration compatible with tox 3 and 4. Define in requires (`tox>=4.2` - better than minversion as is more generic, personally I'd consider minversion deprecated) and let tox 3 die.

**Reply by ppentchev** [reply]
Thanks again. OK, I will consider the possibilities. I had come across min_version some time ago and then I promptly forgot about it... I guess that I may use a combination of both, we'll see.
